sdk/go/common/util/gitutil/git.go

```diff
@@ -255,9 +301,101 @@ func parseAuthURL(remoteURL string) (string, *http.BasicAuth, error) {
 	return u.String(), auth, nil
 }
 
+// sshConfig allows tests to mock SSH key paths.
+type sshConfig struct {
+	path string
+	err  error
+}
+
+// GetKeyPath reads SSH configuration from disk when the receiver is nil. When
+// non-nil it acts as a mock.
+func (c *sshConfig) GetKeyPath(host string) (string, error) {
+	if c == nil {
+		return ssh_config.GetStrict(host, "IdentityFile")
+	}
+	return c.path, c.err
+}
+
+// getSSHPublicKeys reads from the user's SSH configuration and returns public
+// keys for the given host.
+//
+// The `PULUMI_GITSSH_PASSPHRASE` environment variable can be provided if the
+// relevant key is passphrase protected, or (if in an interactive session) the
+// user will be prompted to input a passphrase.
+//
+// TODO: Integrate with GCM when https://github.com/go-git/go-git/issues/490
+// lands.
+//
+// This method handles `~/.ssh/config`, `/etc/host/ssh`, and `Include`
+// directives in the SSH configuration as you would expect.
+func getSSHPublicKeys(user string, host string, sshConfig *sshConfig) (*git_ssh.PublicKeys, error) {
+	privateKeyPath, err := sshConfig.GetKeyPath(host)
+	if err != nil {
+		return nil, err
+	}
+	// Expand tilde (~) if present in the path.
+	privateKeyPath, err = expandHomeDir(privateKeyPath)
+	if err != nil {
+		return nil, err
+	}
+	logging.V(10).Infof("Inferred SSH key '%s' for Git host %s", privateKeyPath, host)
+
+	privateKeyBytes, err := os.ReadFile(privateKeyPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Attempt to load the key. If this is an interactive session and the key
+	// is passphrase-protected we will prompt the user to enter a passphrase.
+	signer, err := ssh.ParsePrivateKey(privateKeyBytes)
+	if errors.As(err, new(*ssh.PassphraseMissingError)) {
+		passphrase := env.GitSSHPassphrase.Value()
+
+		if passphrase == "" && cmdutil.Interactive() {
+			passphrase, err = cmdutil.ReadConsoleNoEcho(
+				fmt.Sprintf("Enter passphrase for SSH key '%s'", privateKeyPath),
+			)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		signer, err = ssh.ParsePrivateKeyWithPassphrase(privateKeyBytes, []byte(passphrase))
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return &git_ssh.PublicKeys{User: user, Signer: signer}, nil
+}
+
+// expandHomeDir expands file paths relative to the user's home directory (~) into absolute paths.
+func expandHomeDir(path string) (string, error) {
+	if len(path) == 0 {
+		return path, nil
+	}
+
+	if path[0] != '~' {
+		return path, nil
+	}
+
+	if len(path) > 1 && path[1] != '/' && path[1] != '\\' {
+		return "", errors.New("cannot expand user-specific home dir")
```

Comment:
worth adding some comments here:

```suggestion
	if path[0] != '~' {
		// Not a "~/foo" path.
		return path, nil
	}

	if len(path) > 1 && path[1] != '/' && path[1] != '\\' {
		// We won't expand "~user"-style paths.
		return "", errors.New("cannot expand user-specific home dir")
```